packages/marshal/src/types.js

```diff
@@ -2,70 +2,317 @@
+ * PassableKeys, then the CopyArray is a PassableKey that can be compared for
+ * equivalence with `sameKey`. Two such CopyArrays are equivalent if their
+ * elements are pairwise equivalent.
  *
- * A Comparable is a Passable in which none of the leaves of the pass-by-copy
- * superstructure are promises. Two Comparables may be compared by
- * for equivalence according to `sameStructure`, which is the strongest
- * equivalence class supported by marshal's distributed object semantics.
+ * A CopyArray as a JS object is a hardened plain array with no holes, no extra
+ * properties, the normal non-enumerable `length` property, and enumerable
+ * data properties from 0 to length-1.
  */
 
 /**
- * @typedef {Comparable} OnlyData
+ * @template {Passable} T
+ * @typedef {{name: string]: T}} CopyRecord
+ * Corresponding PassStyle is "copyRecord".
  *
- * A Comparable is OnlyData when its pass-by-copy superstructure has no leaves,
- * i.e., when all the leaves of the data structure tree are primitive data
- * types or empty composites.
+ * The semantics of CopyRecords is modeled on Records from the Records and
+ * Tuples proposal.
+ * Property names can only be strings. A CopyRecord consists only of
+ * a set of (property-name, value) pairs, where each property name
+ * can appear at most once. It is therefore a single-valued mapping
+ * from property-name to value. CopyRecords enumerate their properties
+ * according to the sort order of their property names.
+ *
+ * CopyArray and CopyRecord are parameterized over various "deep" constraints
+ * explained below. For example, iff the values of a CopyRecord are
+ * PassableKeys, the CopyRecord is PassableKeys and can be compared for
+ * equivalence based on the pairwise equivalence of their values.
+ * Iff the values of a CopyRecord are sortable, then the CopyRecord is
+ * sortable. Two sortable CopyRecords with the same property names are
+ * compared by lexicographic comparison ordered by the sorted order of
+ * their property names.
  */
 
 /**
- * @typedef {OnlyData} PureData
+ * @typedef {Object} CopyError
+ * Corresponding PassStyle is "copyError".
+ *
+ * TODO What I'd like to say in these types in that only a CopyError
+ * without an `errorId` field is a Comparable.
  *
- * An OnlyData value is PureData when it contains no hidden mutable state,
- * e.g., when none of its pass-by-copy composite data objects are proxies. This
- * cannot be determined by inspection. It can only be achieved by trusted
- * construction. A PureData value cannot be used as a communications channel,
- * and can therefore be safely shared with subgraphs that should not be able
- * to communicate with each other.
+ * ***TODO WARNING*** Since CopyErrors have no analog amount Records and
+ * Tuples, perhaps a CopyError in general should only be a Comparable but
+ * not a PassableKey. Or perhaps not even a Comparable! In that case we
+ * wouldn't need to make a special case for `errorId`.
```

Comment:
This is an interesting and curious point... I have never even considered a reason for comparing errors.  What would the motivation be?

Reply:
Also it sounds like our current versions of errors are both fairly "basic", and we both agree that a zoo of error types will eventually be defined.  I think that's sufficient to think in terms of that being a TODO to be worked out for all of us in the future and back-burner it.  This is good enough for now and I'm happy to mirror its structure on Goblins' end.